patchwork-networking/src/main/java/net/minecraftforge/fml/network/NetworkRegistry.java

```diff
@@ -0,0 +1,389 @@
+	 *
+	 * @see FMLHandshakeMessages.S2CModList
+	 * @see FMLHandshakeMessages.C2SModListReply
+	 */
+	static Map<Identifier, Pair<String, Boolean>> buildChannelVersionsForListPing() {
+		return instances.entrySet().stream().
+			map(p -> Pair.of(p.getKey(), Pair.of(p.getValue().getNetworkProtocolVersion(), p.getValue().tryClientVersionOnServer(ABSENT)))).
+			filter(p -> !p.getLeft().getNamespace().equals("fml")).
+			collect(Collectors.toMap(Pair::getLeft, Pair::getRight));
+	}
+
+	static List<String> listRejectedVanillaMods(BiFunction<NetworkInstance, String, Boolean> testFunction) {
+		final List<Pair<Identifier, Boolean>> results = instances.values().stream().
+			map(ni -> {
+				final String incomingVersion = ACCEPTVANILLA;
+				final boolean test = testFunction.apply(ni, incomingVersion);
+				LOGGER.debug(NETREGISTRY, "Channel '{}' : Vanilla acceptance test: {}", ni.getChannelName(), test ? "ACCEPTED" : "REJECTED");
+				return Pair.of(ni.getChannelName(), test);
+			}).filter(p -> !p.getRight()).collect(Collectors.toList());
+
+		if (!results.isEmpty()) {
+			LOGGER.error(NETREGISTRY, "Channels [{}] rejected vanilla connections",
+				results.stream().map(Pair::getLeft).map(Object::toString).collect(Collectors.joining(",")));
+			return results.stream().map(Pair::getLeft).map(Object::toString).collect(Collectors.toList());
+		}
+		LOGGER.debug(NETREGISTRY, "Accepting channel list from vanilla");
+		return Collections.emptyList();
+	}
+
+	/**
+	 * Validate the channels from the server on the client. Tests the client predicates against the server
+	 * supplied network protocol version.
+	 *
+	 * @param channels An @{@link Map} of name->version pairs for testing
+	 * @return true if all channels accept themselves
+	 */
+	static boolean validateClientChannels(final Map<Identifier, String> channels) {
+		return validateChannels(channels, "server", NetworkInstance::tryServerVersionOnClient);
+	}
+
+	/**
+	 * Validate the channels from the client on the server. Tests the server predicates against the client
+	 * supplied network protocol version.
+	 * @param channels An @{@link Map} of name->version pairs for testing
+	 * @return true if all channels accept themselves
+	 */
+	static boolean validateServerChannels(final Map<Identifier, String> channels) {
+		return validateChannels(channels, "client", NetworkInstance::tryClientVersionOnServer);
+	}
+
+	/**
+	 * Tests if the map matches with the supplied predicate tester
+	 *
+	 * @param incoming An @{@link Map} of name->version pairs for testing
+	 * @param originName A label for use in logging (where the version pairs came from)
+	 * @param testFunction The test function to use for testing
+	 * @return true if all channels accept themselves
+	 */
+	private static boolean validateChannels(final Map<Identifier, String> incoming, final String originName, BiFunction<NetworkInstance, String, Boolean> testFunction) {
+		final List<Pair<Identifier, Boolean>> results = instances.values().stream().
```

Comment:
Can this be replaced with a simple `for` loop? It doesn't really seem like it's a very sensible lambda.